streams/src/main/java/org/apache/kafka/streams/processor/internals/Task.java

```diff
@@ -26,26 +26,64 @@
 import java.util.Set;
 
 public interface Task {
+    enum State {
+        CREATED, RESTORING, RUNNING, SUSPENDED, CLOSED;
 
-    void initializeMetadata();
+        static void validateTransition(final State oldState, final State newState) {
+            if (oldState == CREATED && (newState == RESTORING || newState == CLOSED)) {
+                return;
+            } else if (oldState == RESTORING && (newState == RUNNING || newState == SUSPENDED)) {
+                return;
+            } else if (oldState == RUNNING && (newState == RESTORING || newState == SUSPENDED)) {
+                return;
+            } else if (oldState == SUSPENDED && (newState == RUNNING || newState == CLOSED)) {
+                return;
+            } else {
+                throw new IllegalStateException("Invalid transition from " + oldState + " to " + newState);
+            }
+        }
+    }
+
+    State state();
+
+    void transitionTo(State newState);
+
+    void initializeIfNeeded();
+
+    void startRunning();
+
+//    void initializeMetadata();
 
     /**
-     * Initialize the task and return {@code true} if the task is ready to run, i.e, it has no state stores
-     * @return true if this task has no state stores that may need restoring.
+     * Initialize the task's stores
      * @throws IllegalStateException If store gets registered after initialized is already finished
      * @throws StreamsException if the store's change log does not contain the partition
      */
-    boolean initializeStateStores();
+//    void initializeStateStores();
+
+//    boolean hasChangelogs();
 
     boolean commitNeeded();
 
-    void initializeTopology();
+//    void initializeTopology();
 
     void commit();
 
+    void suspend();
```

Comment:
How about removing suspend / resume from Task since they are only for StreamTask?

Reply:
I added it as part of an effort to make the task management in general ignorant of the specific type of task. E.g., when the TaskManager calls this method, it's because it entered a rebalance and therefore wants to suspend all processing work. StandbyTasks know that they don't have any processing work, so they can implement this as a no-op, but the TaskManager doesn't need to worry about whether different kinds of tasks have processing work. It can just tell all tasks to suspend processing. Standby tasks can say (completely truthfully) "Ok, I've suspended all active processing (by doing nothing, since I never had any to begin with".

Reply:
Okay, but in TaskManager I saw you still specifically loop over active tasks first then standby tasks :)

Reply:
Aha! That was to invoke the active or standby task creator, I think... Of course we have to know what kind of task we need to construct so we can call the right constructor.